File: go/pgx/src/example_preferred.go
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

// Package example_preferred demonstrates concurrent queries using the DSQL connector pool.
//
// Works with both admin and non-admin users:
//   - Admin users operate in the default "public" schema
//   - Non-admin users operate in a custom "myschema" schema
package example_preferred

import (
	"context"
	"errors"
	"fmt"
	"os"
	"sync"

	"github.com/awslabs/aurora-dsql-connectors/go/pgx/dsql"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

const numConcurrentQueries = 8

func createPool(ctx context.Context, clusterEndpoint, clusterUser string) (*pgxpool.Pool, error) {
	poolCfg, _ := pgxpool.ParseConfig("")
	poolCfg.MaxConns = 10
	poolCfg.MinConns = 2

	// Set search_path on each new connection based on user type
	var schema string
	if clusterUser == "admin" {
		schema = "public"
	} else {
		schema = "myschema"
	}
	poolCfg.AfterConnect = func(ctx context.Context, conn *pgx.Conn) error {
		_, err := conn.Exec(ctx, fmt.Sprintf("SET search_path = %s", pgx.Identifier{schema}.Sanitize()))
		if err != nil {
			return fmt.Errorf("failed to set search_path to %s: %w", schema, err)
		}
		return nil
	}

	return dsql.NewPool(ctx, dsql.Config{
		Host: clusterEndpoint,
		User: clusterUser,
	}, poolCfg)
}

// workerResult holds either a successful result or an error from a worker.
type workerResult struct {
	workerID int
	result   string
	err      error
}

func worker(ctx context.Context, pool *pgxpool.Pool, workerID int) workerResult {
	var result int
	err := pool.QueryRow(ctx, "SELECT $1::int as worker_id", workerID).Scan(&result)
	if err != nil {
		return workerResult{workerID: workerID, err: fmt.Errorf("worker %d error: %w", workerID, err)}
	}
	return workerResult{workerID: workerID, result: fmt.Sprintf("Worker %d result: %d", workerID, result)}
}

// Example demonstrates concurrent queries using the DSQL connector pool.
func Example() error {
	clusterEndpoint := os.Getenv("CLUSTER_ENDPOINT")
	if clusterEndpoint == "" {
		return fmt.Errorf("CLUSTER_ENDPOINT environment variable is not set")
	}
	clusterUser := os.Getenv("CLUSTER_USER")
	if clusterUser == "" {
		clusterUser = "admin"
	}

	ctx := context.Background()

	pool, err := createPool(ctx, clusterEndpoint, clusterUser)
	if err != nil {
		return fmt.Errorf("failed to create pool: %w", err)
	}
	defer pool.Close()

	// Verify connection
	if err := pool.Ping(ctx); err != nil {
		return fmt.Errorf("failed to ping: %w", err)
	}

	// Run concurrent queries using the connection pool
	results := make(chan workerResult, numConcurrentQueries)

	var wg sync.WaitGroup
	for i := 1; i <= numConcurrentQueries; i++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()
			results <- worker(ctx, pool, workerID)
		}(i)
	}

	// Wait for all workers to complete and close the channel
	wg.Wait()
	close(results)

	// Collect results and errors
	var errs []error
	for res := range results {
		if res.err != nil {
			errs = append(errs, res.err)
		} else {
			fmt.Println(res.result)
		}
	}

	// Return combined errors if any occurred
	if len(errs) > 0 {
		return errors.Join(errs...)
	}

	fmt.Println("Connection pool with concurrent connections exercised successfully")
	return nil
}

File: python/asyncpg/src/example_preferred.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

import asyncio
import os

import aurora_dsql_asyncpg as dsql


async def worker_task(pool, worker_id):
    """Simulate concurrent database operations."""

    async with pool.acquire() as conn:
        result = await conn.fetchval("SELECT $1::int", worker_id)
        return result


async def connect_with_pool_concurrent_connections(cluster_user, cluster_endpoint):
    ssl_cert_path = "./root.pem"
    if not os.path.isfile(ssl_cert_path):
        raise FileNotFoundError(f"SSL certificate file not found: {ssl_cert_path}")

    pool_params = {
        "user": cluster_user,
        "host": cluster_endpoint,
        "ssl": "verify-full",
        "sslrootcert": ssl_cert_path,
        "min_size": 5,
        "max_size": 10,
    }

    pool = None
    try:
        pool = await dsql.create_pool(retry=True, **pool_params)

        # Run multiple concurrent workers
        num_workers = 5
        tasks = [worker_task(pool, i) for i in range(num_workers)]
        results = await asyncio.gather(*tasks)
        for result in results:
            print(result)

        async def insert_owner(conn):
            await conn.execute(
                "INSERT INTO owner(name, city, telephone) VALUES($1, $2, $3)",
                "John Doe",
                "Anytown",
                "555-555-1900",
            )

        await pool.run_transaction(insert_owner)
    finally:
        if pool is not None:
            await pool.close()


async def main():
    try:
        cluster_user = os.environ.get("CLUSTER_USER", None)
        assert cluster_user is not None, "CLUSTER_USER environment variable is not set"

        cluster_endpoint = os.environ.get("CLUSTER_ENDPOINT", None)
        assert cluster_endpoint is not None, "CLUSTER_ENDPOINT environment variable is not set"

        await connect_with_pool_concurrent_connections(cluster_user, cluster_endpoint)

    finally:
        pass

    print("Concurrent pool operations completed successfully")


if __name__ == "__main__":
    asyncio.run(main())
